### PR TITLE
DRAFT: Add django-q2 as a task service

### PR DESCRIPTION
**Pull request by rachaelcodes**
## Context

We need to schedule, and run some repeated Django tasks. Ideally we would do this in a way that allows us to view the success or otherwise of those tasks quickly, and with minimal extra overhead. I've chosen [django-q2](https://django-q2.readthedocs.io/en/master/) as a [fairly well-maintained and popular](https://snyk.io/advisor/python/django-q2) package that integrates well with Django, including allowing us to view and edit tasks in Django Admin.

Screenshot 2024-07-11 at 11 16 00

## Changes proposed in this pull request

Adds `django-q2` to the Django app with the Postgres database as the message broker. Its `qcluster` task monitoring service has been set up as a new Docker app. I've moved the 'delete expired data' task to a new `tasks` directory.

__This is a draft PR__ Still to add:
* tests
* cron capability
* more developed documentation

## Guidance to review

Try running this locally. To run `qcluster`, either run the `django-q` Docker app or run it from the shell in `/redbox/django_app` with `poetry run python manage.py qcluster`.

Visit Django Admin and take a look at the new Django Q2 pages. Try setting up a new scheduled task from the Django admin form. You can run the 'delete expired data' task by setting `func` to `redbox_app.redbox_core.tasks.delete_expired_data.task`. 

@252afh do you think the infrastructure for this would be fairly straightforward to set up? 

## Relevant links
https://technologyprogramme.atlassian.net/browse/REDBOX-460

## Things to check

- [ ] I have added any new ENV vars in all deployed environments
- [ ] I have tested any code added or changed
- [ ] I have run integration tests

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -26,6 +26,22 @@ services:
...
+    build:
+      context: .
+      dockerfile: ./django_app/django_q/Dockerfile
+    networks:
+      - redbox-app-network
+    env_file:
+      - .env
+    volumes:
+      - ./django_app:/app/django_q/
+    depends_on:
```
Comment (brunns) [inline docker-compose.yml:40]: We may need to depend on `core-api` for your task.

File: django_app/redbox_app/redbox_core/tasks/README.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+In the longer term, if we end up with many more tasks, we might want to go for a different broker or even task runner, but this will involve setting up more infrastructure.
+
+## Scheduling tasks
+
+Tasks can be scheduled, managed and monitored from Django Admin at: `/admin/django_q/`.
+
+Task functions can be from anywhere in the django codebase; to organise them, there is a `tasks` directory inside `redbox_core`.
+
+To schedule a new task, visit `/admin/django_q/schedule/` and 'Add Scheduled Task'.
```
Comment (brunns) [inline django_app/redbox_app/redbox_core/tasks/README.md:20]: We could write a migration to pre-populate this task.

**Review by brunns — APPROVED** [review]
Very nice.

File: django_app/redbox_app/redbox_core/tasks/README.md (hunk 2 above)
Comment (rachaelcodes) [reply django_app/redbox_app/redbox_core/tasks/README.md:20]: ooh good idea

**Review by gecBurton — COMMENTED** [review]
all looks very sensible: a minimal amount of code commensurate with the size/significance of the task. Please could you briefly detail why other options such as celery were ruled out

**Comment by rachaelcodes** [comment]
Closing in favour of going serverless: #776 🌍

**Pull request closed without merging**